stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -605,44 +673,73 @@ impl NakamotoBlock {
             return None;
         }
 
-        if coinbase_positions.len() > 1 {
-            // has more than one coinbase
+        if coinbase_positions.len() > 1 || tenure_change_positions.len() > 1 {
+            // never valid to have more than one of each
             return Some(false);
         }
 
         if coinbase_positions.len() == 1 && tenure_change_positions.len() == 0 {
-            // has a coinbase but no tenure change
+            // coinbase unaccompanied by a tenure change
             return Some(false);
         }
 
-        if coinbase_positions.len() == 0 && tenure_change_positions.len() > 0 {
-            // has tenure-changes but no coinbase
-            return Some(false);
-        }
+        if coinbase_positions.len() == 0 && tenure_change_positions.len() == 1 {
+            // this is possibly a block with a tenure-extend transaction.
+            // It must be the first tx
+            if tenure_change_positions != vec![0] {
+                // wrong position
+                return Some(false);
+            }
+
+            // must be a non-sortition-triggered tenure change
+            let TransactionPayload::TenureChange(tc_payload) = &self.txs[0].payload else {
+                // this transaction is not a tenure change
+                // (should be unreachable)
+                return Some(false);
+            };
```

Comment:
Can you DRY some of this checking up with what's going on in the next check as well? It seems like in all cases, the tenure_change_transaction _must_ be at position 0, and you want to access the payload in any event so you can check `.expects_sortition`: